### DISCUSSION TITLE
Duplicate CSV transaction is not detected FIDI v0.9.16

### DISCUSSION THREAD

**Discussion by cubagithub** [firefly-iii/firefly-iii#6483] [Firefly III] [answered]
### Support guidelines

- [X] I've read the [support guidelines](https://github.com/firefly-iii/firefly-iii/blob/main/.github/support.md)

### I've found a bug and checked that ...

- [X] ... [the documentation](https://docs.firefly-iii.org/) does not mention anything about my problem
- [X] ... there are no open or closed issues that are related to my problem
- [X] ... it's [definitely me, not you](https://github.com/firefly-iii/firefly-iii/blob/main/.github/its_you_not_me.md)

### Description
![image](https://user-images.githubusercontent.com/34730232/193439847-e492b1ec-5067-451d-ba17-67ad6c2b4431.png)

This is how the transaction looks in the CSV file:
```
Date,Description,Credit,Debit,Balance
12/02/2022,DELIVEROO - Visa Purchase - Receipt 109397In <REDACTED> Date 10 Feb 2022 Card <REDACTED>,,-36.39,202.67
```

The original transaction was imported months ago using the deprecated `fireflyiii/csv-importer` container - here's a snippet of the import JSON config:
```json
    "column-roles": [
        "date-transaction",
        "description",
        "amount_credit",
        "amount_debit",
        "_ignore"
    ],
    "column-do-mapping": [
        false,
        false,
        false,
        false,
        false
    ]
```

The duplicate transaction was imported today using the new `fireflyiii/data-importer` container - here's a snippet of its import JSON config:
```json
    "roles": [
        "date_transaction",
        "description",
        "amount_credit",
        "amount_debit",
        "_ignore"
    ],
    "do_mapping": [false, false, false, false, false]
```

### Expected behaviour

Duplicate transaction with identical line data should be skipped during import.

### Debug information

- [Debug JSON view of original transaction in Firefly](https://gist.github.com/cubagithub/e930e5c342802e1256d9e7106357cf92)
- [Debug JSON view of duplicated transaction in Firefly](https://gist.github.com/cubagithub/c1aad1b8ab9b816cb3a2b87202cd8233)
- [Debug output of the import from console](https://gist.github.com/cubagithub/930dd76180e1cb64a5910821c189b0f3)

### Steps to reproduce

1. copy CSV transactions file with previously imported rows to container's `/import` volume
2. execute `docker exec -it firefly-import php artisan importer:auto-import /import`

### Additional info

Firefly III Version [5.7.12] (docker)
Firefly III data importer v0.9.16 (docker)

**Comment by JC5** [comment, marked as answer]
I can see the hash is different between the old transaction and the new one, even though I can't see differences myself Without knowing what was submitted by the CSV importer originally, I can't tell why that would be the case. Ergo, there isn't much I can do about this. 

Much has changed between the old CSV importer and the new data importer.

**Comment by cubagithub** [comment]
Any way I can pull the debug log from an old import job?

**Comment by JC5** [comment]
I'm afraid not, unless you stored the old Docker container logs.

**Comment by cubagithub** [comment]
Do the hashes matter? I'm using "duplicate_detection_method": "classic".

**Reply by JC5** [reply]
Yes, the classic duplicate detection uses the import_hash_v2

**Comment by cubagithub** [comment]
Since this is a discussion thread now, may I ask some things about `fireflyiii/data-importer` container behaviour that I couldn't answer from the docs...

1. will data-importer sleep/stop after a period of inactivity, and need to be started again with `docker start`? I recall this was required for the old importer.
2. does data-importer look at its `/import` volume path on startup and process and found csv+json combos? Or do you need to intentionally run the `docker exec -it firefly-import php artisan importer:auto-import /import` regardless?

FYI I wasn't able to determine what is causing the duplicate detections during import, so I've just resolved this with a local python script to find the most recent transaction for each account and trim earlier rows from my CSV import files.

**Reply by JC5** [reply]
Sure thing!

1. It will keep running in the background.
2. You need to call it intentionally, it will not check automatically.

**Reply by cubagithub** [reply]
I have some doubts around this behaviour, if you wouldn't mind clarifying please. 

This https://docs.firefly-iii.org/data-importer/usage/command_line/#run-automatic-imports suggests that the container should be run on-demand for an automatic import, which I assume would terminate after completion, right?

I'm using a persistent container as per https://docs.firefly-iii.org/data-importer/install/docker/ with docker-compose (see yaml below) but it appears it's terminating itself after some idle time (see output below).

```yaml
  firefly-import:
    image: fireflyiii/data-importer:latest
    container_name: firefly-import
    environment:
      ...
      - IMPORT_DIR_WHITELIST=/import
    volumes:
      - <REDACTED>:/import
    depends_on:
      - firefly
    ...
    restart: unless-stopped

  firefly:
    image: fireflyiii/core:latest
    container_name: firefly
    env_file: .env
    ...
    restart: always
```

```bash
  $ docker ps -a
  IMAGE,COMMAND,CREATED,STATUS
  
  fireflyiii/data-importer:latest		
  2 days ago
  Exited (0) 40 hours ago
  
  fireflyiii/core:latest
  2 days ago
  Up 10 hours (healthy)   
```

In this case should I be copying my CSVs to the `/import` path and starting the container? Or do I need to start it and then run this command: https://docs.firefly-iii.org/data-importer/usage/command_line/#importing-a-directory ? Or is something not working as intended?

**Reply by JC5** [reply]
> but it appears it's terminating itself after some idle time (see output below).

It is not designed to do so. This could be a bug of some sorts, and I am curious about any logs you may have.

> In this case should I be copying my CSVs to the /import path and starting the container?

It depends on you, both would work.

1. Start the container using [this command](https://docs.firefly-iii.org/data-importer/usage/command_line/#run-automatic-imports), so it will auto import and then exit.
2. Keep the container running, refresh the files in the `/import` dir, and then run [this command](https://docs.firefly-iii.org/data-importer/usage/command_line/#importing-a-directory).
